substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/config/ConfigurationKind.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2019, 2019, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.configure.config;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public final class ConfigurationKind {
+    private static Map<ConfigurationMemberKind, Map<ConfigurationAccessKind, ConfigurationKind>> cache = new HashMap<>();
```

Comment:
This cache seems like a memory leak as it stays here forever. Also, it takes a bit of time to understand how it works due to the map-in-a-map structure.

To avoid such confusion, I would propose to use a two-dimensional array here as both keys are enums. It will take less space, and it will immediately be obvious what it is.

Reply:
If the `cache` remains, I think it needs to be all `ConcurrentHashMap` since it is filled lazily.

Reply:
Switched to a 2D-array